server/src/main/java/org/elasticsearch/index/fielddata/LeafFieldData.java

```diff
@@ -37,4 +41,26 @@
      */
     SortedBinaryDocValues getBytesValues();
 
+    /**
+     * Return a value fetcher for this leaf implementation.
+     */
+    default DocValueFetcher.Leaf getLeafValueFetcher(DocValueFormat format) {
+        SortedBinaryDocValues values = getBytesValues();
+        return new DocValueFetcher.Leaf() {
```

Comment:
I was about to ask if we are good considering this the default impl, but I double checked the old logic from FetchDocValuesPhase and it seems to match the index field data class hierarchy, so +1. One concern I have, and I have been wondering for a while, why don't we have a base class for all the Sorted*DocValues given that all of the advanceExact and docValueCount impl are exactly the same? nextValue also is very similar besides its return type.

Reply:
I think some of the `SortedNumericDocValues` comes from Lucene and it is sort of "theirs". I think `LeafFieldData` is "our" layer on top of it. I'm not 100% sure I'd have designed it all this way, but it is how we're using it now. I think.

Reply:
right, SortedNumericDocvalues comes from Lucene, while the others (double, binary etc) we added. I guess that we needed only a subset of the methods that lucene has on its impls as they all extend DocValuesIterator.

It is probably complicated to change the classes hierarchy then.

Reply:
We probably could make a shared base class for the ones that are "ours" though. I feel like we should leave the interface because we're often adapting from the double one to the long one and back. So really there are, like, two common implementations, one that wraps and one that counts.

But I think that is a change for another time.